### PR TITLE
Enhance TFTrainer.save_model()

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Currently, `TFTrainer.save_model()` raises errors if the model is not `TFPreTrainedModel` . However `Trainer` works fine with `torch.nn.modules.Module`. 

This is a step to make TFTrainer work with usual `tf.keras.models.Model` models. The idea (from @sgugger) is that a user is building their own models that work like ours (e.g., return the loss as the first output) and can train them with Trainer.

Furthermore, a SavedModel is also saved using `tf.saved_model.save()`.

I tried to avoid duplicated code (check and create output directory before saving), and therefore there is a new method `save_tf_model()` in `modeling_tf_utils`, which is used in `trainer_tf.py`.

For @jplu and @sgugger .

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -301,6 +301,31 @@ def load_tf_weights(model, resolved_archive_file):
...
+        return
+    os.makedirs(save_directory, exist_ok=True)
+
+    # If we save using the predefined names, we can load using `from_pretrained` if ``model`` is an instance of class `TFPreTrainedModel`.
+    output_model_file = os.path.join(save_directory, TF2_WEIGHTS_NAME)
+    model.save_weights(output_model_file)
+    logger.info("Model weights saved in {}".format(output_model_file))
+
+    # Also save to SavedModel format.
+    tf.saved_model.save(model, export_dir=save_directory)
```
Comment (jplu) [inline src/transformers/modeling_tf_utils.py:325]: No, saving a `saved_model` must be done **only** in the trainer, and should not happen when calling `.save_pretrained()`.

**Review by jplu — CHANGES_REQUESTED** [review]
Thanks for the PR!

Just one comment, you just have to modify the `TFTrainer` not the `TFPretrainedModel` because a saved model has to be created only after a training. 

Basically the method `TFTrainer.save_model(self, output_dir)` must become:

```
def save_model(self, output_dir):
    output_dir = output_dir if output_dir is not None else self.args.output_dir
    logger.info("Saving model in {}".format(output_dir))
    if isinstance(self.model, TFPreTrainedModel):
        self.model.save_pretrained(output_dir)
    else:
        self.model.save_weights(output_dir)
    tf.saved_model.save(self.model, output_dir)
```

And `TFPreTrainedModel.save_pretrained(self, save_directory)` must stay as it is.

File: src/transformers/modeling_tf_utils.py (hunk 1 above)
Comment (chiapas) [reply src/transformers/modeling_tf_utils.py:325]: Hi, 

The modified `TFPreTrainedModel.save_pretrained` is actually equivalent to its previous form - if we don't add `tf.saved_model.save` part. The test failed only after I add this new part today.

Maybe I missed some requirement I don't know. I will try further to see if I can get something working by avoiding using tf.saved_model.save() inside a call to `save_pretrained()`.

Your code snippet is good, but I just tried to avoid the duplicated code like

```
    if os.path.isfile(save_directory):
        logger.error("Provided path ({}) should be a directory, not a file".format(save_directory))
        return
    os.makedirs(save_directory, exist_ok=True)
```

Comment (jplu) [reply src/transformers/modeling_tf_utils.py:325]: The tests fails because now the saving part is outside. Make it back inside the class and the tests will pass. Having few duplicated lines of code is not a big deal.

Comment (chiapas) [reply src/transformers/modeling_tf_utils.py]: It turns out that it works after I removed

```
        # Also save to SavedModel format.
        tf.saved_model.save(self.model, export_dir=output_dir)
        logger.info("SavedModel saved in {}".format(output_dir))
```

to `TFTrainer.save_model()`. Would this current version acceptable, @jplu ?

Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Yep, this is normal, it is because the saved model TF function could not find the proper serialization for the parameters.

Can you just put back as it was, please?

Comment (chiapas) [reply src/transformers/modeling_tf_utils.py]: OK, I could do that. But would you mind to give me some more explanation? The current version moves `self.save_weights` outside `TFPreTrainedModel`, but it should be the same thing. And current version passes the tests, so there shouldn't be any problem.

Sorry to bother, but it is good to know what's going on here. Thanks for your patience.

Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: No worries it is ok :)

It is mostly because I'm currently reworking a lot of things of TF side including the way we save/load the weights. So the less changes on this part the less updates I will do later 😄

Comment (chiapas) [reply src/transformers/modeling_tf_utils.py]: OK, I got it now :) I will rollback to its previous form, no problem. Thanks again.

**Comment by jplu** [comment]
Closing in favor of #7619
